### ISSUE TITLE
Luigi memory consumption causing spark jobs to crash

### ISSUE THREAD

**Issue by cqnguyen** [spotify/luigi#963]
Hey guys,

Just discovered that when running luigi with spark, found my jobs were failing due to OOM errors. When running the same job without luigi in spark, the job succeeds successfully.

Spark jobs with large data sets fail when running via luigi. Small spark jobs relative to cluster size succeed when running through luigi. 

Cheers!
Cory

**Comment by Tarrasch** [comment]
We're not using spark for luigi at Spotify. So I can't help this time. :)

I think you'll have to dig through the code and see if something strange is going on. Btw, are the OOM exceptions locally or on the cluster?

**Comment by cqnguyen** [comment]
The OOM are occurring on the cluster

**Comment by Tarrasch** [comment]
Aha I see. I have no idea how luigi could be related to that

**Comment by hellais** [comment]
I have not used luigi with spark myself yet, but have intention in doing so, hence this issue is quite relevant to my interest.

@cqnguyen could you provide a code example of how you use `SparkSubmitTask`, `PySparkTask` or `luigi.contrib.pyspark_runner` (I am assuming you are using one of those).

Also how much memory do the executors in your cluster have?

From looking at the code the most critical parts of the code are:

* https://github.com/spotify/luigi/blob/master/luigi/contrib/pyspark_runner.py#L43 (I would add before the pickle.load something that prints out the pickle file so that you can inspect if it's too big and why)

*  https://github.com/spotify/luigi/blob/master/luigi/contrib/spark.py#L366, I am not sure exactly how addFiles works, but it may be worthwhile looking in here too

Furthermore to help in debugging you should also print out the available memory on the executor by adding something like:

```
print(sc._conf.get('spark.executor.memory'))
```

in your `main`.

**Comment by hellais** [comment]
Ha, I just ran into this issue.

Here is some debug log output in case somebody can help me look into this issue:



```
# There is insufficient memory for the Java Runtime Environment to continue.
# Native memory allocation (malloc) failed to allocate 1431830528 bytes for committing reserved memor
y.
# Possible reasons:
#   The system is out of physical RAM or swap space
#   In 32 bit mode, the process size limit was hit
# Possible solutions:
#   Reduce memory load on the system
#   Increase physical memory or swap space
#   Check if swap backing store is full
#   Use 64 bit Java on a 64 bit OS
#   Decrease Java heap size (-Xmx/-Xms)
#   Decrease number of Java threads
#   Decrease Java thread stack sizes (-Xss)
#   Set larger code cache with -XX:ReservedCodeCacheSize=
# This output file may be truncated or incomplete.
#
#  Out of Memory Error (os_linux.cpp:2745), pid=8799, tid=139924741879552
#
# JRE version:  (7.0_71-b14) (build )
# Java VM: Java HotSpot(TM) 64-Bit Server VM (24.71-b01 mixed mode linux-amd64 compressed oops)
# Core dump written. Default location: /home/hadoop/ooni-pipeline-ng/core or core.8799
#

---------------  T H R E A D  ---------------

Current thread (0x00007f42bc016000):  JavaThread "Unknown thread" [_thread_in_vm, id=8804, stack(0x00007f42c478f000,0x00007f42c4890000)]

Stack: [0x00007f42c478f000,0x00007f42c4890000],  sp=0x00007f42c488e230,  free space=1020k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x9a0a1a]  VMError::report_and_die()+0x2ea
V  [libjvm.so+0x4974bb]  report_vm_out_of_memory(char const*, int, unsigned long, char const*)+0x9b
V  [libjvm.so+0x81e4de]  os::Linux::commit_memory_impl(char*, unsigned long, bool)+0xfe
V  [libjvm.so+0x81ea29]  os::pd_commit_memory(char*, unsigned long, unsigned long, bool)+0x29
V  [libjvm.so+0x8187da]  os::commit_memory(char*, unsigned long, unsigned long, bool)+0x2a
V  [libjvm.so+0x88b2e3]  PSVirtualSpace::expand_by(unsigned long)+0x53
V  [libjvm.so+0x879f73]  PSOldGen::initialize(ReservedSpace, unsigned long, char const*, int)+0x103
V  [libjvm.so+0x29a593]  AdjoiningGenerations::AdjoiningGenerations(ReservedSpace, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long)+0x3e3
V  [libjvm.so+0x83f325]  ParallelScavengeHeap::initialize()+0x4d5
V  [libjvm.so+0x97295a]  Universe::initialize_heap()+0xca
V  [libjvm.so+0x973b29]  universe_init()+0x79

[snip]

/proc/meminfo:
MemTotal:       15407772 kB
MemFree:         1157972 kB
MemAvailable:    1240180 kB
Buffers:            3324 kB
Cached:           229348 kB
SwapCached:            0 kB
Active:         13968580 kB
Inactive:          35192 kB
Active(anon):   13771120 kB
Inactive(anon):       20 kB
Active(file):     197460 kB
Inactive(file):    35172 kB
Unevictable:           0 kB
Mlocked:               0 kB
SwapTotal:             0 kB
SwapFree:              0 kB
Dirty:              5152 kB
Writeback:             0 kB
AnonPages:      13771196 kB
Mapped:            40692 kB
Shmem:                40 kB
Slab:              68052 kB
SReclaimable:      26784 kB
SUnreclaim:        41268 kB
KernelStack:       15176 kB
PageTables:        57576 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:     7703884 kB
Committed_AS:   50820668 kB
VmallocTotal:   34359738367 kB
VmallocUsed:       37740 kB
VmallocChunk:   34359641671 kB
AnonHugePages:         0 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
DirectMap4k:    15736832 kB
DirectMap2M:           0 kB

CPU:total 4 (16 cores per cpu, 2 threads per core) family 6 model 62 stepping 4, cmov, cx8, fxsr, mmx
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 62
model name      : Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
stepping        : 4
microcode       : 0x415
cpu MHz         : 2500.112
cache size      : 25600 KB
physical id     : 0
siblings        : 4
core id         : 0
cpu cores       : 1
apicid          : 0
initial apicid  : 0
fpu             : yes
```

So from the looks of it I'm running out of 16GB of RAM. Is this normal?

This is the spark task I am running: https://github.com/TheTorProject/ooni-pipeline-ng/blob/master/pipeline/batch/spark_apps.py#L134

**Comment by Tarrasch** [comment]
I continue to have a hard time seeing how luigi could relate to this. First, luigi is just running a given command line when shelling out to spark (I assume), second, the failure is on the cluster, the jobs running there don't know about luigi, right? Last, I've heard over and over again that spark is still a bit beta and that it's suddenly starts to hog memory.

Now with this said, the problem can still be with luigi of course, but it wouldn't be the first place I would look at when debugging this.

**Comment by hellais** [comment]
Yes you are probably right, I probably would not have opened a ticket on luigi if I hadn't already found one present here.

By doing some finer tuning of the amount of workers I used I was able to run my spark queries just fine on the same cluster.

As a rule of thumb I would say use 1 worker * the number of nodes in your cluster (I previously had a number of workers that was more than double that).

The luigi code specific to spark does indeed support your claim that it doesn't do much apart from packaging the dependencies and running the spark-submit command with the dependencies + the task script.

I think this can be closed.
